lib/TPP/ConvertLinalgToTpp.cpp

```diff
@@ -66,8 +66,6 @@ struct ConvertGenericOpToTpp : public OpRewritePattern<linalg::GenericOp> {
 
   LogicalResult matchAndRewrite(linalg::GenericOp linalgOp,
                                 PatternRewriter &rewriter) const override {
-    if (!linalgOp.hasBufferSemantics())
```

Comment:
Do we want to restrict this rewrite (later) to only operate at tensor level?

Reply:
Imo, lowering is simple and generic enough that we can support both tensors and memrefs.
Probably many transformation and optimizations will be restricted to tensor but that should be fine.

Reply:
I'm concerned that this will allow us to "choose" at which level we want to lower, and then end up with two semantics in one implementation. In the past we ended up with code that has `hasBufferSemantics` checks everywhere, and then it gets hard to know what works with what (`N:M` problem).

We only want to lower once. Tensor, or Memref. So, if it's already lowered at Tensor, we should have no more linalg left overs at Memref level. If we do, it was introduced by some other pass and perhaps we shouldn't play with it?

Regardless, in compilers, it's always safer to block unintentional usage than allow "because it might be useful later". Rule of thumb is that it rarely is and when it is, we can't use the same code.

Reply:
>  In the past we ended up with code that has hasBufferSemantics checks everywhere

Tpp ops exist in both tensor and memref worlds, we have to perform this validation anyway if a pass expects to work only on tensors or memrefs.

Restricting lowering only to tensors seems a bit artificial. But if supporting lkowering both for tensors and memref adds significant maintenance overhead then it would be reasonable to constrain it.

> We only want to lower once.

I agree, hopefully we won't need double lowering in the default pipeline, and I guess most of our future passes will require tpp on tensors.